# Long Short-Term Memory (LSTM) with PyTorch + Lightning

This notebook will demonstrate how to create, optimize and make predictions using the Long Short-Term Memory (LSTM) network.

Specifically, we will implement the Long Short-Term Memory unit seen below, that predicts sequential data to predict the value of two different companies.

<img src='images/lstm_image.001.png' style='width: 720px'>

The training data (below) consist of stock prices for two different companies, **Company A** and **Company B**. The goal is to use the data from the first **4** days to predict what the price will be on the **5th** day. If we look closely at the data, we'll see that the only differences in the prices occur on Day **1** and Day **5**. So the LSTM has to remember what happened on Day **1** in order to predict what will happen on Day **5**.


<img src='images/company_a_data.png' style='width: 360px'> <img src='images/company_b_data.png' style='width: 360px'>


-----

## Import the modules that will do all the work

The very first thing we need to do is load a bunch of Python modules. These modules give us extra functionality to create a Long Short-Term Memory (LSTM) neural network and optimize the neural network's parameters.

In [1]:
import torch
import torch.nn as nn
import torch.functional as F
from torch.optim import Adam

import lightning as L
from torch.utils.data import TensorDataset, DataLoader

import matplotlib.pyplot as plt
import seaborn as sns

-----

## Build a Long Short-Term Memory unit by hand using PyTorch + Lightning

A Long Short-Term Memory (LSTM) unit is a type of neural network, and that means we need to create a new class. To make it easy to train the LSTM, this class will inherit from `LightningModule` and we'll create the following methods:
- `__init__()` to initialize the Weights and Biases and keep track of a few other house keeping things.
- `lstm_unit()` to do the LSTM math.
- `forward()` to make a forward pass through the unrolled LSTM.
- `configure_optimizers()` to configure the optimizer.
- `training_step()` to pass the training data to `forward()`, calculate the loss and to keep track of the loss values in a log file.


In [6]:
class LSTMbyHand(L.LightningModule):
    def __init__(self):
        super().__init__()

        L.seed_everything(seed=30)
        
        mean = torch.tensor(0.0)
        std = torch.tensor(1.0)

        self.wlr1 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.wlr2 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.blr1 = nn.Parameter(torch.tensor(0.0), requires_grad=True)

        self.wpr1 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.wpr2 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.bpr1 = nn.Parameter(torch.tensor(0.0), requires_grad=True)

        self.wp1 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.wp2 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.bp1 = nn.Parameter(torch.tensor(0.0), requires_grad=True)

        self.wo1 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.wo2 = nn.Parameter(torch.normal(mean=mean, std=std), requires_grad=True)
        self.bo1 = nn.Parameter(torch.tensor(0.0), requires_grad=True)

    def lstm_unit(self, input_value, long_memory, short_memory):
        long_remember_percent = torch.sigmoid((short_memory * self.wlr1) + 
                                             (input_value * self.wlr2) + 
                                             self.blr1)

        potential_remember_percent = torch.sigmoid((short_memory * self.wpr1) +
                                                   (input_value * self.wpr2) +
                                                   self.bpr1)
        
        potential_memory = torch.tanh((short_memory * self.wp1) + 
                                      (input_value * self.wp2) +
                                      self.bp1)

        updated_long_memory = ((long_memory * long_remember_percent)+
                               (potential_remember_percent * potential_memory))

        output_percent = torch.sigmoid((short_memory * self.wo1) +
                                       (input_value * self.wo2))

        updated_short_memory = torch.tanh(updated_long_memory) * output_percent

        return ([updated_long_memory, updated_short_memory])


    def forward(self, X):
        long_memory = 0
        short_memory = 0

        day1 = X[0]
        day2 = X[1]
        day3 = X[2]
        day4 = X[3]

        long_memory, short_memory = self.lstm_unit(day1, long_memory, short_memory)
        long_memory, short_memory = self.lstm_unit(day2, long_memory, short_memory)
        long_memory, short_memory = self.lstm_unit(day3, long_memory, short_memory)
        long_memory, short_memory = self.lstm_unit(day4, long_memory, short_memory)

        return short_memory
    
    def configure_optimizers(self):
        return Adam(self.parameters())
    
    def training_step(self, batch, batch_idx):
        input_i, label_i = batch
        output_i = self.forward(input_i[0])
        loss = (label_i - output_i)**2

        self.log('train_loss', loss)

        if label_i == 0:
            self.log('out_0', output_i)
        else:
            self.log('out_1', output_i)


        return loss

Once we have created the class that defines an LST, we can use it to create a model and print out the randomly initialized Weights and Biases. Then, we'll see what those random Weights and Biases predict for **Company A** and **Company B**. If they are good predictions, then we're done! However, the chances of getting good predictions from random values is very small.

In [7]:
model = LSTMbyHand()

print('Paramaters before optimization')
for name, parameter in model.named_parameters():
    print(f'{name} : {parameter.data}')

print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

Seed set to 30


Paramaters before optimization
wlr1 : 0.4704992175102234
wlr2 : 1.6563351154327393
blr1 : 0.0
wpr1 : 0.5152892470359802
wpr2 : -0.2744353711605072
bpr1 : 0.0
wp1 : -2.260566473007202
wp2 : 1.227964162826538
bp1 : 0.0
wo1 : 0.7928141355514526
wo2 : 0.6231480836868286
bo1 : 0.0
Company A: Observed = 0, Predicted = tensor(0.2914, grad_fn=<MulBackward0>)
Company B: Observed = 1, Predicted = tensor(0.2860, grad_fn=<MulBackward0>)


With the unoptimized parameters, the predicted value for **Company A**, **0.2914**, is quite far from the observed value, **0**. On the other hand, the predicted value for **Company B**, **0.2860**, is terrible, because it is relatively far from the observed value, **1**. So, that means we need to train the LSTM.

-----

## Train the LSTM unit and use Lightning and TensorBoard to evaluate

Since we are using **Lightning** training, training the LSTM we created by hand is pretty easy. All we have to do is create the training data and put it into a `DataLoader`.

In [8]:
inputs = torch.tensor([[0., 0.5, 0.25, 1.], [1., 0.5, 0.25, 1.]])
labels = torch.tensor([0., 1.])

dataset = TensorDataset(inputs, labels)
dataloader = DataLoader(dataset)

We then create a **Lightning Trainer**, `L.Trainer`, and fit it to the training data. **NOTE:** We are starting with **2000** epochs. This may be enough to successfully optimize all of the parameters, but it might not.

In [9]:
trainer = L.Trainer(max_epochs=2000)
trainer.fit(model, train_dataloaders=dataloader)

print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name         | Type | Params
--------------------------------------
  | other params | n/a  | 12    
--------------------------------------
12        Trainable params
0         Non-trainable params
12        Total params
0.000     Total estimated model params size (MB)
/home/gabryel/development/ai/pytorch/basics/neuralnet/venv/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.
/home/gabryel/development/ai/pytorch/basics/neuralnet/venv/lib/python3.10/site-packages/lightning/pytorch/loops/fit_loop.py:298: The number of training batches (2) is smaller than th

Epoch 1999: 100%|██████████| 2/2 [00:00<00:00,  9.34it/s, v_num=13]

`Trainer.fit` stopped: `max_epochs=2000` reached.


Epoch 1999: 100%|██████████| 2/2 [00:00<00:00,  5.55it/s, v_num=13]
Company A: Observed = 0, Predicted = tensor(0.4669, grad_fn=<MulBackward0>)
Company B: Observed = 1, Predicted = tensor(0.5329, grad_fn=<MulBackward0>)


Unfortunately, these predictions are terrible. So, it seems like we'll have to do more training. However, it would be great if we could be confident that more training will actually improve the predictions. If not, we can spare ourselves a lot of time, and potentially money, and just give up. So, before we dive into more training, let's look at the loss values and predictions that we saved in logfiles with **TensorBoard**. **TensorBoard** will graph everything that we logged during training, making it easier to see if things are headed in the right direction or not.

Below are the graphs of **loss** (`train_loss`), the predictions for **Company A** (`out_0`), and the predictions for **Company B** (`out_1`). Remember for **Company A**, we want to predict **0** and for **Company B**, we want to predict **1**.

<img src='images/trainloss_2000_epochs.png' style='width: 360px'> <img src='images/out_0_2000_epochs.png' style='width: 360px'> <img src='images/out_1_2000_epochs.png' style='width: 360px'>

If we look at the **loss** (`train_loss`), we see that it is going down, which is good, but it still has further to go. When we look at the predictions for **Company A** (`out_0`), we see that they started out pretty good, close to **0**, but then got really bad early on in training, shooting all the way up to **0.5**, but are starting to get smaller. In contrast, when we look at the predictions for **Company B** (`out_1`), we see that they started out really bad, close to **0**, but have been getting better ever since and look like they could continue to get better if we kept training.

In summary, the graphs seem to suggest that if we continued training our model, the predictions would improve.

## Adding More Epochs without Straing Over

Using Lightning, we can pick up where we left off training without having to start from scratch. This is because when we train with **Lightning**, it creates __checkpoint__ files that keep track of the Weights and Biases as they change. As a result, all we have to do to piuck up where we left off is tell the `Trainer` where the checkpoint files are located. Since we want to add **1000** more epochs to the training, we indicate **3000**.

In [10]:
path_to_best_checkpoint = trainer.checkpoint_callback.best_model_path

trainer = L.Trainer(max_epochs=3000)
trainer.fit(model, train_dataloaders=dataloader, ckpt_path=path_to_best_checkpoint)

print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Restoring states from the checkpoint path at /home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_13/checkpoints/epoch=1999-step=4000.ckpt
/home/gabryel/development/ai/pytorch/basics/neuralnet/venv/lib/python3.10/site-packages/lightning/pytorch/callbacks/model_checkpoint.py:361: The dirpath has changed from '/home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_13/checkpoints' to '/home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_14/checkpoints', therefore `best_model_score`, `kth_best_model_path`, `kth_value`, `last_model_path` and `best_k_models` won't be reloaded. Only `best_model_path` will be reloaded.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name         | Type | Params
--------------------------------------
  | other params | n/a  | 12    
-----------

Epoch 2003:   0%|          | 0/2 [00:00<?, ?it/s, v_num=14]        

Epoch 2999: 100%|██████████| 2/2 [00:00<00:00, 17.02it/s, v_num=14]

`Trainer.fit` stopped: `max_epochs=3000` reached.


Epoch 2999: 100%|██████████| 2/2 [00:00<00:00, 10.37it/s, v_num=14]
Company A: Observed = 0, Predicted = tensor(0.3573, grad_fn=<MulBackward0>)
Company B: Observed = 1, Predicted = tensor(0.6169, grad_fn=<MulBackward0>)


<img src='images/trainloss_3000_epochs.png' style='width: 360px'> <img src='images/out_0_3000_epochs.png' style='width: 360px'> <img src='images/out_1_3000_epochs.png' style='width: 360px'>

The graphs are much better than before. The orange lines in each graph represents the values we logged during the extra **1000** epochs. The **loss** is getting smaller and the predictions for both companies are improving. However, because it looks like there is even more room for improvement, let's add **2000** more epochs to the training.

In [11]:
path_to_best_checkpoint = trainer.checkpoint_callback.best_model_path

trainer = L.Trainer(max_epochs=5000)
trainer.fit(model, train_dataloaders=dataloader, ckpt_path=path_to_best_checkpoint)

print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Restoring states from the checkpoint path at /home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_14/checkpoints/epoch=2999-step=6000.ckpt
/home/gabryel/development/ai/pytorch/basics/neuralnet/venv/lib/python3.10/site-packages/lightning/pytorch/callbacks/model_checkpoint.py:361: The dirpath has changed from '/home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_14/checkpoints' to '/home/gabryel/development/ai/pytorch/basics/neuralnet/lightning_logs/version_15/checkpoints', therefore `best_model_score`, `kth_best_model_path`, `kth_value`, `last_model_path` and `best_k_models` won't be reloaded. Only `best_model_path` will be reloaded.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name         | Type | Params
--------------------------------------
  | other params | n/a  | 12    
-----------

Epoch 4999: 100%|██████████| 2/2 [00:00<00:00, 39.21it/s, v_num=15]

`Trainer.fit` stopped: `max_epochs=5000` reached.


Epoch 4999: 100%|██████████| 2/2 [00:00<00:00, 26.35it/s, v_num=15]
Company A: Observed = 0, Predicted = tensor(0.0017, grad_fn=<MulBackward0>)
Company B: Observed = 1, Predicted = tensor(0.9567, grad_fn=<MulBackward0>)


<img src='images/trainloss_5000_epochs.png' style='width:360px'> <img src='images/out_0_5000_epochs.png' style='width:360px'> <img src='images/out_1_5000_epochs.png' style='width:360px'>

After **5000** epochs, the graphs have converged. The prediction for **Company A** is super close to **0**, which is exactly what we want, and the prediction for **Company B** is close to **1**, which is also what we want.


The blue lines show how things changed when we added an additional **2000** epochs to the training, for a total of **5000** epochs. Now we see that the **loss** (`train_loss`) and the predictions for each company appear to be tapering off, suggesting that adding more epochs may not improve the predictions much, so we're done with the training.


Finally, let's print out the final estimates for the Weights and Biases.

In [12]:
print('Paramaters after optimization')
for name, parameter in model.named_parameters():
    print(f'{name} : {parameter.data}')

Paramaters after optimization
wlr1 : 2.799105167388916
wlr2 : 3.443237543106079
blr1 : 1.8347482681274414
wpr1 : 2.3876936435699463
wpr2 : 0.23879866302013397
bpr1 : 0.3615899384021759
wp1 : 1.1882497072219849
wp2 : 1.7998336553573608
bp1 : -0.6124131679534912
wo1 : 2.4440248012542725
wo2 : 1.557745337486267
bo1 : 0.0


-----

## Using and optimizing the PyTorch LSTM using nn.LSTM()

PyTorch provides an easier way to produce an LSTM model using `nn.LSTM()`, this allows us to simplify the `__init__()` function and the `forward()` function. The other big difference is that this time, we're not going to try and recreate the parameter values we used in the previous LSTM model.

In [13]:
class LightningLSTM(L.LightningModule):
    def __init__(self):
        super().__init__()

        L.seed_everything(seed=30)
        
        self.lstm = nn.LSTM(input_size=1, hidden_size=1)

    def forward(self, X):
        X_T = X.view(len(X), 1)

        lstm_out, temp = self.lstm(X_T)

        prediction = lstm_out[-1]
        
        return prediction
    
    def configure_optimizers(self):
        return Adam(self.parameters(), lr=0.1)
    
    def training_step(self, batch, batch_idx):
        input_i, label_i = batch
        output_i = self.forward(input_i[0])
        loss = (label_i - output_i)**2

        self.log('train_loss', loss)

        if label_i == 0:
            self.log('out_0', output_i)
        else:
            self.log('out_1', output_i)

        return loss

Now let's create the model and print out the initial Weights, Biases, and predictions

In [14]:
model = LightningLSTM()

print('Paramaters before optimization')
for name, parameter in model.named_parameters():
    print(f'{name} : {parameter.data}')

print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

Seed set to 30


Paramaters before optimization
lstm.weight_ih_l0 : tensor([[ 0.8015],
        [ 0.4927],
        [-0.0568],
        [ 0.7475]])
lstm.weight_hh_l0 : tensor([[0.4805],
        [0.5680],
        [0.7892],
        [0.2477]])
lstm.bias_ih_l0 : tensor([-0.1448,  0.6842,  0.4909,  0.2363])
lstm.bias_hh_l0 : tensor([ 0.7765, -0.1745, -0.6504, -0.3147])
Company A: Observed = 0, Predicted = tensor([-0.3027], grad_fn=<SelectBackward0>)
Company B: Observed = 1, Predicted = tensor([-0.3323], grad_fn=<SelectBackward0>)


As expected, the predictions are bad, so we will train the model. However, because we've increased the learning rate to 0.1, we only need to train for **300** epochs.

In [15]:
trainer = L.Trainer(max_epochs=300, log_every_n_steps=2)
trainer.fit(model, train_dataloaders=dataloader)

print('Paramaters before optimization')
for name, parameter in model.named_parameters():
    print(f'{name} : {parameter.data}')
    
print('Company A: Observed = 0, Predicted =', model(torch.tensor([0., 0.5, 0.25, 1.]).detach()))
print('Company B: Observed = 1, Predicted =', model(torch.tensor([1., 0.5, 0.25, 1.]).detach()))

GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name | Type | Params
------------------------------
0 | lstm | LSTM | 16    
------------------------------
16        Trainable params
0         Non-trainable params
16        Total params
0.000     Total estimated model params size (MB)
/home/gabryel/development/ai/pytorch/basics/neuralnet/venv/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.


Epoch 299: 100%|██████████| 2/2 [00:00<00:00, 18.75it/s, v_num=16]

`Trainer.fit` stopped: `max_epochs=300` reached.


Epoch 299: 100%|██████████| 2/2 [00:00<00:00, 11.88it/s, v_num=16]
Paramaters before optimization
lstm.weight_ih_l0 : tensor([[2.1643],
        [0.6354],
        [1.3501],
        [1.2649]])
lstm.weight_hh_l0 : tensor([[2.9409],
        [3.2667],
        [2.5271],
        [2.5610]])
lstm.bias_ih_l0 : tensor([-0.8395,  1.1769,  0.3222,  0.7485])
lstm.bias_hh_l0 : tensor([ 0.0818,  0.3182, -0.8190,  0.1976])
Company A: Observed = 0, Predicted = tensor([4.7270e-05], grad_fn=<SelectBackward0>)
Company B: Observed = 1, Predicted = tensor([0.9859], grad_fn=<SelectBackward0>)


As we can see, after just **300** epochs, the LSTM is making great predictions. The prediction for **Company A** is close to the observed value **0** and the prediction for **Company B** is close to the observed value **1**.

<img src='images/trainloss_3.png' style='width:360px'> <img src='images/out_0_3.png' style='width:360px'> <img src='images/out_1_3.png' style='width:360px'>

In all three graphs, the **loss** (`train_loss`) and the predictions for **Company A** (`out_0`) and **Company B** (`out_1`) started to taper off after **500** steps, or just **250** epochs, suggesting that adding more epochs may not improve the predictions much.